### PR TITLE
Add functions to view and define regions/synonyms for timeseries data

### LINKED ISSUE #28: Integrate `nodes` from MESSAGE-scheme `Scenario` set and the `timeseries` data
At the moment, defining a new element in the set `node` of a MESSAGE-scheme `ixmp.Scenario` does not add an entry to the table `IAMC_NODES`, meaning that one cannot directly use that node in the `add_timeseries()` command.

### PR DESCRIPTION
**Pull request by danielhuppmann**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- [x] Description in RELEASE_NOTES.md Added

# Description

This PR adds a number of functions to view and define new regions to be used in the IAMC-style timeseries.

Closes #28

### CODE REVIEW

**Comment by zikolach** [comment]
Looks good to me. @danielhuppmann did you include ixmp.jar built after merging `ixmp_source` branch into master? (just to be sure https://github.com/iiasa/ixmp_source/pull/179 changes are included)

File: tests/test_regions.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    test_mp.add_region('foo', 'bar', 'World')
+    obs = test_mp.regions()
+    assert all(obs[obs['region'] == 'foo'] == ['foo', None, 'World', 'bar'])
+
+
+def test_add_region_synonym(test_mp):
+    test_mp.add_region_synomym('foo2', 'foo')
+    regions = test_mp.regions()
+    obs = regions[regions.region.isin(['foo', 'foo2'])].reset_index(drop=True)
+    print(obs)
```
Comment (gidden) [inline tests/test_regions.py]: rm print please!

**Comment by gidden** [comment]
lgtm! with release notes, stickler, and print removal should be good to go

**Comment by danielhuppmann** [comment]
>  @danielhuppmann did you include ixmp.jar built after merging `ixmp_source` branch into master? (just to be sure [iiasa/ixmp_source#179](https://github.com/iiasa/ixmp_source/pull/179) changes are included)

Yes, I just rebuilt and force-pushed over the previous commit to make sure that the `jar` file is based on the latest `ixmp_source` version.

File: ixmp/core.py [hunk 2]
```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
...
+        the synonym in the region column, it will be converted to `mapped_to`.
+
+        Parameters
+        ----------
+        region : str
+            Name of the region synonym.
+        mapped_to : str
+            Name of the region to which the synonym should be mapped.
+        """
+        self._jobj.addNodeSynonym(mapped_to, region)
```
Comment (khaeru) [inline ixmp/core.py]: What happens if `mapped_to` doesn't exist? Should probably raise `ValueError`. Specify in docstring and test.

File: ixmp/core.py [hunk 3]
```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
...
+        *Before adding a region, please use `regions()` and check whether the
+        region already exists with a different spelling.
+        If so, use `add_region_synonym()` instead.
+
+        Parameters
+        ----------
+        region : str
+            Name of the region.
+        hierarchy : str
+            Hierarchy level of the region (e.g., country, R11, basin)
```
Comment (khaeru) [inline ixmp/core.py]: Does the hierarchy level need to exist? Maybe say (I'm guessing) "If the given level does not exist, it is created."

A question as an aside: how is the hierarchy of levels set? E.g. where/how is it specified that an 'R11' contains 1 or more 'country'; or that a particular country is in a particular R11?

File: ixmp/core.py [hunk 4]
```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
         """
         self._jobj.addUnitToDB(unit, comment)
 
+    def regions(self):
+        """Return all regions defined for the IAMC-style timeseries format
+        including known synonyms.
+
+        Returns
+        -------
+        pd.Dataframe
```
Comment (khaeru) [inline ixmp/core.py]:
```
:class:`pandas.DataFrame`
```
as in, e.g. the docstring for `TimeSeries.timeseries()`.

File: ixmp/core.py (hunk 3 above)
Comment (danielhuppmann) [reply ixmp/core.py]: The hierarchy level and the mapping to regions is currently purely informative and there are no validation steps - the database will accept any values and any mapping (as long as the `parent` region exists in the `region` column).

File: ixmp/core.py (hunk 4 above)
Comment (danielhuppmann) [reply ixmp/core.py]: done

File: ixmp/core.py [hunk 5]
```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
...
+            Name of the region.
+        hierarchy : str
+            Hierarchy level of the region (e.g., country, R11, basin)
+        parent : str, default 'World'
+            Assign a 'parent' region.
+        """
+        self._jobj.addNode(region, parent, hierarchy)
+
+    def add_region_synomym(self, region, mapped_to):
+        """Define a synomym for a `region`. When adding timeseries data using
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply ixmp/core.py]: done

File: ixmp/core.py [hunk 6]
```diff
@@ -250,6 +250,53 @@ def add_unit(self, unit, comment='None'):
...
+            n, p, h = (r.getName(), r.getParent(), r.getHierarchy())
+            lst.extend([(n, None, p, h)])
+            lst.extend([(s, n, p, h) for s in (r.getSynonyms() or [])])
+        region = pd.DataFrame(lst)
+        region.columns = ['region', 'mapped_to', 'parent', 'hierarchy']
+        return region
+
+    def add_region(self, region, hierarchy, parent='World'):
+        """Define a region including a hierarchy level and a 'parent' region.
+        *Before adding a region, please use `regions()` and check whether the
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply ixmp/core.py]: done

**Comment by danielhuppmann** [comment]
merging as approved by @gidden, @zikolach and @khaeru  to make CI for https://github.com/iiasa/message_ix/pull/142 pass

**Pull request merged**